### PR TITLE
[SPARK-47422][SQL] Support collated strings in array operations

### PR DESCRIPTION
**Pull request by nikolamand-db**
### What changes were proposed in this pull request?

Collations need to be properly supported in following array operations but currently yield unexpected results: `ArraysOverlap`, `ArrayDistinct`, `ArrayUnion`, `ArrayIntersect`, `ArrayExcept`. Example query:
```
select array_contains(array('aaa' collate utf8_binary_lcase), 'AAA' collate utf8_binary_lcase)
```
We would expect the result of query to be true.

### Why are the changes needed?

To enable collation support in array operations.

### Does this PR introduce _any_ user-facing change?

Yes, with these changes listed array operations return expected results when applied on arrays of collated strings.

### How was this patch tested?

Added test to collections suite and updated collations golden files.

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CollectionExpressionsSuite.scala [hunk 1]
```diff
@@ -2821,4 +2821,88 @@ class CollectionExpressionsSuite
       a, Literal(5), Literal.create("q", StringType)), Seq("b", "a", "c", null, "q")
     )
   }
+
+  test("Array operations on arrays of collated strings") {
```
Comment (dbatomic) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CollectionExpressionsSuite.scala:2825]: IMO, golden file tests already cover most of these cases.
Secondly, general agreement is to keep collation tests grouping in either CollationSuite or CollationExpressionSuite. Let's keep it that way.

Besides that LGTM.

**Review by dbatomic — APPROVED** [review]

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CollectionExpressionsSuite.scala (hunk 1 above)
Comment (cloud-fan) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CollectionExpressionsSuite.scala:2825]: +1

File: sql/core/src/test/resources/sql-tests/inputs/collations.sql [hunk 2]
```diff
@@ -70,3 +70,29 @@ select array_position(ARRAY('aaa' collate utf8_binary_lcase, 'bbb' collate utf8_
 -- utility
 select nullif('aaa' COLLATE utf8_binary_lcase, 'AAA' COLLATE utf8_binary_lcase);
 select least('aaa' COLLATE utf8_binary_lcase, 'AAA' collate utf8_binary_lcase, 'a' collate utf8_binary_lcase);
+
+-- array operations
+select arrays_overlap(array('aaa' collate utf8_binary_lcase), array('AAA' collate utf8_binary_lcase));
```
[the remark this answers is not in the document]
Comment (cloud-fan) [reply sql/core/src/test/resources/sql-tests/inputs/collations.sql:75]: +1, we should write scala unit tests for test coverage. end-to-end is mostly testing the integration part.

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**